runtime/src/main/cpp/MemorySharedRefs.hpp

```diff
@@ -37,13 +53,40 @@ class BackRefFromAssociatedObject {
  public:
   void initAndAddRef(ObjHeader* obj);
 
+  // Error if refCount is zero and it's called from the wrong worker with non-frozen obj_.
+  template <ErrorHandlingPolicy errorHandlingPolicy>
   void addRef();
+  void addRefOrTerminate() {
+    addRef<ErrorHandlingPolicy::kTerminate>();
+  }
+  void addRefOrThrow() {
+    addRef<ErrorHandlingPolicy::kThrowException>();
+  }
 
+  // Error if called from the wrong worker with non-frozen obj_.
+  template <ErrorHandlingPolicy errorHandlingPolicy>
   bool tryAddRef();
+  bool tryAddRefOrTerminate() {
+    return tryAddRef<ErrorHandlingPolicy::kTerminate>();
+  }
+  bool tryAddRefOrThrow() {
+    return tryAddRef<ErrorHandlingPolicy::kThrowException>();
+  }
 
   void releaseRef();
 
+  // Error if called from the wrong worker with non-frozen obj_.
+  template <ErrorHandlingPolicy errorHandlingPolicy>
   ObjHeader* ref() const;
+  ObjHeader* refOrTerminate() const {
```

Comment:
Again, new functions where parameterized one already exist. Just extra layer and mind bloating.
Is  `refOrThrow()` simpler than `ref<kThrow>()`? You may consider shorter names for parameter though.